### PR TITLE
Merge Payment Requests into 2.0

### PR DESCRIPTION
**Pull request by GSadee**
| Q               | A
|-----------------|-----
| Branch?         | 2.0
| Bug fix?        | no
| New feature?    |yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: UPGRADE-2.0.md [hunk 1]
```diff
@@ -6,6 +6,9 @@
     * The `sylius_default.bus` and `sylius_event.bus` configuration options were removed. Use `sylius.command_bus` and
       `sylius.event_bus` for commands and events respectively.
 
+* SyliusStateMachineAbstraction:
+    * The `sylius_state_machine_abstraction.default_adapter` option has been changed from `winzou_state_machine` to `symfony_workflow`.
+
```
Comment (NoResponseMate) [inline UPGRADE-2.0.md:11]: This doesn't seem to be related

File: config/routes/test_cached_payum/routing.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+sylius_test_plugin_main:
+    path: /test/main
+    controller: Symfony\Bundle\FrameworkBundle\Controller\TemplateController::templateAction
+    defaults:
+        template: "@SyliusTestPlugin/main.html.twig"
```
Comment (NoResponseMate) [inline config/routes/test_cached_payum/routing.yaml:5]: Is this needed?

File: config/services_test_cached_payum.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+imports:
+    - { resource: "services_test_cached.yaml" }
+
+services:
+    ##
+    # Prioritize Payum by tagging services with a higher priority (or lower for tagged_iterator).
+    ##
+
```
Comment (NoResponseMate) [inline config/services_test_cached_payum.yaml]:
```suggestion
    # Prioritize Payum by tagging services with a higher priority (or lower for tagged_iterator).
```

File: features/admin/product/managing_products/sorting_products_within_a_taxon_by_position.feature [hunk 4]
```diff
@@ -80,7 +80,7 @@ Feature: Sorting listed products from a taxon by position
         And the 7th product on this page should be named "Pug Master"
         And this product should be at position 16
 
-    @no-api @ui @mink:chromedriver
+    @no-api @tod-ui @mink:chromedriver
```
Comment (NoResponseMate) [inline features/admin/product/managing_products/sorting_products_within_a_taxon_by_position.feature]: What happened here?

File: phpstan-baseline.neon [hunk 5]
```diff
@@ -1100,6 +1100,26 @@ parameters:
 			count: 1
 			path: src/Sylius/Bundle/CoreBundle/Order/Checker/OrderPromotionsIntegrityChecker.php
 
+		-
+			message: "#^Call to method toBinary\\(\\) on an unknown class Symfony\\\\Component\\\\Uid\\\\Uuid\\.$#"
```
Comment (NoResponseMate) [inline phpstan-baseline.neon]: This seems wrong, but we'll have to regenerate the baseline either way so 🤷

File: src/Sylius/Behat/Context/Api/Shop/PaymentRequestContext.php [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    public function myPaymentRequestShouldBeCancelled(string $action, PaymentMethodInterface $paymentMethod, string $state): void
+    {
+        $request = $this->getRequestForPaymentRequestWithAction($action);
+        Assert::notNull($request, sprintf('Payment request with action %s not found', $action));
+
+        $this->client->executeCustomRequest($request);
+        $response = $this->client->getLastResponse();
+
+        Assert::same($this->responseChecker->getValue($response, 'action'), $action, sprintf('Payment request should have action %s', $action));
+        Assert::true(str_contains($this->responseChecker->getValue($response, 'method'), $paymentMethod->getCode()), sprintf('Payment request should have payment method %s', $paymentMethod->getCode()));
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Shop/PaymentRequestContext.php]:
```suggestion
        Assert::contains($this->responseChecker->getValue($response, 'method'), $paymentMethod->getCode(), sprintf('Payment request should have payment method %s', $paymentMethod->getCode()));
```

File: src/Sylius/Behat/Resources/config/services/contexts/api/shop.xml [hunk 7]
```diff
@@ -161,6 +161,16 @@
...
         </service>
 
+        <service id="sylius.behat.context.api.shop.payment_request" class="Sylius\Behat\Context\Api\Shop\PaymentRequestContext">
+            <argument type="service" id="sylius.behat.shared_storage" />
+            <argument type="service" id="sylius.behat.api_platform_client.shop" />
+            <argument type="service" id="Sylius\Behat\Client\ResponseCheckerInterface" />
+            <argument type="service" id="sylius.behat.request_factory" />
+            <argument type="service" id="sylius.repository.payment_request" />
+            <argument type="service" id="sylius.behat.page.shop.payment_request.payment_method_notify" />
+            <argument type="service" id="test.client" />
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Resources/config/services/contexts/api/shop.xml]:
```suggestion
```

File: src/Sylius/Bundle/AdminBundle/templates/order/show/content/sections/payments/item/actions/list_payment_requests.html.twig [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{% from '@SyliusAdmin/shared/helper/icon.html.twig' import icon %}
+
+{% set payment = hookable_metadata.context.payment %}
+
+<a title="details" href="{{ path('sylius_admin_payment_request_index', {'paymentId': payment.id}) }}" class="btn btn-icon" data-bs-toggle="tooltip" data-bs-title="{{ 'sylius.ui.list_payment_requests'|trans }}" {{ sylius_test_html_attribute('list_payment_requests') }}>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/templates/order/show/content/sections/payments/item/actions/list_payment_requests.html.twig]:
```suggestion
<a title="details" href="{{ path('sylius_admin_payment_request_index', {'paymentId': payment.id}) }}" class="btn btn-icon" data-bs-toggle="tooltip" data-bs-title="{{ 'sylius.ui.list_payment_requests'|trans }}" {{ sylius_test_html_attribute('list-payment-requests') }}>
```

File: src/Sylius/Behat/Page/Admin/Payment/IndexPage.php [hunk 9]
```diff
@@ -48,6 +48,17 @@ public function showOrderPageForNthPayment(int $position): void
...
 
+    public function showPaymentRequestOfNthPayment(int $position): void
+    {
+        $tableAccessor = $this->getTableAccessor();
+        $table = $this->getElement('table');
+
+        $row = $tableAccessor->getRowsWithFields($table, [])[$position];
+        $field = $tableAccessor->getFieldFromRow($table, $row, 'actions');
+
+        $field->find('css', '[data-test-show-action="List payment requests"]')->click();
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Page/Admin/Payment/IndexPage.php:59]:
```suggestion
        $field->find('css', '[data-test-list-payment-requests]')->click();
```

File: src/Sylius/Bundle/ApiBundle/Command/Payment/AddPaymentRequest.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+use Sylius\Bundle\ApiBundle\Attribute\OrderTokenValueAware;
+use Sylius\Bundle\ApiBundle\Command\IriToIdentifierConversionAwareInterface;
+
+/** @experimental */
+#[OrderTokenValueAware]
+class AddPaymentRequest implements IriToIdentifierConversionAwareInterface
+{
+    public function __construct(
+        public readonly string $orderTokenValue,
+        public readonly int|string $paymentId,
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Command/Payment/AddPaymentRequest.php]:
```suggestion
        public readonly mixed $paymentId,
```
?

File: src/Sylius/Bundle/ApiBundle/EventSubscriber/PaymentRequestEventSubscriber.php [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        ];
+    }
+
+    public function postWrite(ViewEvent $event): void
+    {
+        $paymentRequest = $event->getControllerResult();
+
+        if (!$paymentRequest instanceof PaymentRequestInterface) {
+            return;
+        }
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/EventSubscriber/PaymentRequestEventSubscriber.php]:
```suggestion
        $paymentRequest = $event->getControllerResult();
        if (!$paymentRequest instanceof PaymentRequestInterface) {
            return;
        }
```

File: src/Sylius/Bundle/ApiBundle/EventSubscriber/PaymentRequestEventSubscriber.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+        $method = $event->getRequest()->getMethod();
+
+        if (in_array($method, [Request::METHOD_POST, Request::METHOD_PUT, Request::METHOD_PATCH], true)) {
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/EventSubscriber/PaymentRequestEventSubscriber.php]:
```suggestion
        if (in_array($event->getRequest()->getMethod(), [Request::METHOD_POST, Request::METHOD_PUT, Request::METHOD_PATCH], true)) {
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PaymentRequest.xml [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        <operations>
+            <operation
+                name="sylius_api_admin_payment_request_get"
+                class="ApiPlatform\Metadata\Get"
+                uriTemplate="/admin/payment-requests/{hash}"
+            >
+                <normalizationContext>
+                    <values>
+                        <value>sylius:admin:payment_request:show</value>
+                    </values>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/admin/PaymentRequest.xml]: Seems it's also missing the `groups` node 💃

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/ChosenPaymentRequestActionEligibilityValidator.php [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        $paymentMethod = $this->paymentMethodRepository->findOneBy(['code' => $addPaymentRequest->paymentMethodCode]);
+
+        if ($paymentMethod === null) {
+            return null;
+        }
+
+        $gatewayConfig = $paymentMethod->getGatewayConfig();
+        if (null === $gatewayConfig) {
+            return null;
+        }
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/ChosenPaymentRequestActionEligibilityValidator.php]:
```suggestion
        $paymentMethod = $this->paymentMethodRepository->findOneBy(['code' => $addPaymentRequest->paymentMethodCode]);
        if ($paymentMethod?->getGatewayConfig() === null) {
            return null;
        }
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/ChosenPaymentRequestActionEligibilityValidator.php [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+        $factoryName = $this->gatewayFactoryNameProvider->provide($paymentMethod);
+        $commandProvider = $this->gatewayFactoryCommandProvider->getCommandProvider($factoryName);
+
+        return $commandProvider ?? null;
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/ChosenPaymentRequestActionEligibilityValidator.php]:
```suggestion

        return $this->gatewayFactoryCommandProvider->getCommandProvider($factoryName);
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/PaymentRequest.xml [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+-->
+
+<serializer xmlns="http://symfony.com/schema/dic/serializer-mapping"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/serializer-mapping https://symfony.com/schema/dic/serializer-mapping/serializer-mapping-1.0.xsd"
+>
+    <class name="Sylius\Component\Payment\Model\PaymentRequest">
+        <attribute name="id" ignore="true" />
+
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/PaymentRequest.xml:21]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Checker/UpdatePaymentRequestEligibilityChecker.php [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+use Sylius\Component\Payment\Model\PaymentRequestInterface;
+
+final class UpdatePaymentRequestEligibilityChecker implements UpdatePaymentRequestEligibilityCheckerInterface
+{
+    public function isEligible(PaymentRequestInterface $paymentRequest): bool
+    {
+        return in_array($paymentRequest->getState(), [PaymentRequestInterface::STATE_NEW, PaymentRequestInterface::STATE_PROCESSING]);
+    }
+}
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Checker/UpdatePaymentRequestEligibilityChecker.php]: Guess we could think about using the newly introduced `FinalizedPaymentRequestChecker` in place of this one, as it's pretty much the same logic, only the other way around. Less code necessary to customize for the end users as well.

File: src/Sylius/Bundle/CoreBundle/OrderPay/Provider/NoPaymentPayResponseProvider.php [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    ) {
+    }
+
+    public function getResponse(
+        RequestConfiguration $requestConfiguration,
+        OrderInterface $order
+    ): Response {
+        $url = $this->orderPayFinalUrlProvider->getUrl(null);
+
+        return new RedirectResponse($url);
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/OrderPay/Provider/NoPaymentPayResponseProvider.php]:
```suggestion
        return new RedirectResponse($this->orderPayFinalUrlProvider->getUrl(null));
```

File: src/Sylius/Bundle/CoreBundle/OrderPay/Provider/NoPaymentPayResponseProvider.php [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    public function supports(
+        RequestConfiguration $requestConfiguration,
+        OrderInterface $order
+    ): bool {
+        $payment = $this->paymentToPayResolver->getLastPayment($order);
+
+        return null === $payment;
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/OrderPay/Provider/NoPaymentPayResponseProvider.php]:
```suggestion
        return null === $this->paymentToPayResolver->getLastPayment($order);
```

File: src/Sylius/Bundle/CoreBundle/OrderPay/Provider/Offline/StatusHttpResponseProvider.php [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+    public function getResponse(
+        RequestConfiguration $requestConfiguration,
+        PaymentRequestInterface $paymentRequest,
+    ): Response {
+        // Force null payment to go to the thank you page
+        $finalUrl = $this->finalUrlProvider->getUrl(null);
+
+        return new RedirectResponse($finalUrl);
+    }
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/OrderPay/Provider/Offline/StatusHttpResponseProvider.php]:
```suggestion

    // Forces null payments to go to the thank you page
    public function getResponse(
        RequestConfiguration $requestConfiguration,
        PaymentRequestInterface $paymentRequest,
    ): Response {
        return new RedirectResponse($this->finalUrlProvider->getUrl(null));
    }
```

File: src/Sylius/Bundle/CoreBundle/OrderPay/Resolver/PaymentToPayResolverInterface.php [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Sylius\Bundle\CoreBundle\OrderPay\Resolver;
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+
+/** @experimental */
+interface PaymentToPayResolverInterface
+{
+    public function getLastPayment(OrderInterface $order): ?PaymentInterface;
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/OrderPay/Resolver/PaymentToPayResolverInterface.php]:
```suggestion
    public function getPayment(OrderInterface $order): ?PaymentInterface;
```

File: src/Sylius/Bundle/PaymentBundle/Canceller/PaymentRequestCanceller.php [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+                $this->stateMachine->apply(
+                    $paymentRequest,
+                    PaymentRequestTransitions::GRAPH,
+                    PaymentRequestTransitions::TRANSITION_CANCEL,
+                );
+
+                $this->objectManager->persist($paymentRequest);
+                $this->objectManager->flush();
+            }
+        }
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/PaymentBundle/Canceller/PaymentRequestCanceller.php]:
```suggestion
            }
        }

        $this->objectManager->flush();
```
Flushing in a foreach is a bad idea

File: src/Sylius/Bundle/PaymentBundle/Checker/FinalizedPaymentRequestChecker.php [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+    public function isFinal(PaymentRequestInterface $paymentRequest): bool
+    {
+        $state = $paymentRequest->getState();
+        $nextTransition = $this->stateMachine->getTransitionFromState(
+            $paymentRequest,
+            PaymentRequestTransitions::GRAPH,
+            $state,
+        );
+
+        return null === $nextTransition;
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/PaymentBundle/Checker/FinalizedPaymentRequestChecker.php]:
```suggestion
        return null === $this->stateMachine->getTransitionFromState(
            $paymentRequest,
            PaymentRequestTransitions::GRAPH,
            $paymentRequest->getState(),
        );
```

File: src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRequestRepository.php [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+        $paymentRequests = $this->createQueryBuilder('o')
+            ->innerJoin('o.payment', 'payment')
+            ->innerJoin('o.method', 'method')
+            ->where('o != :paymentRequest')
+            ->andWhere('o.action = :action')
+            ->andWhere('o.payment = :payment')
+            ->andWhere('o.method = :method')
+            // Symfony\Bridge\Doctrine\Types\UuidType has signature changes between sf 5.4 and 6.4
+            // We are forced to use type: "uuid" to use both Sf versions
+            ->setParameter('paymentRequest', $paymentRequest->getHash(), 'uuid')
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRequestRepository.php]: 5.4 went away, so I guess this can be resolved 🤔

File: src/Sylius/Bundle/PaymentBundle/Form/Type/PaymentMethodType.php [hunk 25]
```diff
@@ -37,10 +55,40 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
...
         ;
     }
 
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'validation_groups' => function (FormInterface $form): array {
+                $data = $form->getData();
+
+                return $this->paymentMethodGroupsGenerator->__invoke($data);
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/PaymentBundle/Form/Type/PaymentMethodType.php]:
```suggestion
                return $this->paymentMethodGroupsGenerator->__invoke($form->getData());
```

File: src/Sylius/Bundle/PaymentBundle/Generator/GatewayConfigNameGenerator.php [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\PaymentBundle\Generator;
+
+use Sylius\Component\Payment\Model\PaymentMethodInterface;
+
+final readonly class GatewayConfigNameGenerator implements GatewayConfigNameGeneratorInterface
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/PaymentBundle/Generator/GatewayConfigNameGenerator.php]:
```suggestion
final readonly class GatewayNameGenerator implements GatewayNameGeneratorInterface
```

File: src/Sylius/Bundle/PaymentBundle/Provider/AbstractServiceProvider.php [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+namespace Sylius\Bundle\PaymentBundle\Provider;
+
+use Sylius\Bundle\PaymentBundle\Exception\PaymentRequestNotSupportedException;
+use Sylius\Bundle\ResourceBundle\Controller\RequestConfiguration;
+use Sylius\Component\Payment\Model\PaymentRequestInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Contracts\Service\ServiceProviderInterface;
+
+/** @experimental */
+abstract class AbstractServiceProvider implements ServiceProviderAwareProviderInterface
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/PaymentBundle/Provider/AbstractServiceProvider.php:23]: We need to think about simplifying these, and maybe split the logic into a few different services, as it is it's pretty magical and it's easy to get lost

File: src/Sylius/Bundle/ApiBundle/Command/Payment/AddPaymentRequest.php [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Command\Payment;
+
+use Sylius\Bundle\ApiBundle\Attribute\OrderTokenValueAware;
+use Sylius\Bundle\ApiBundle\Command\IriToIdentifierConversionAwareInterface;
+
+/** @experimental */
+#[OrderTokenValueAware]
+class AddPaymentRequest implements IriToIdentifierConversionAwareInterface
```
Comment (Prometee) [inline src/Sylius/Bundle/ApiBundle/Command/Payment/AddPaymentRequest.php:21]:
```suggestion
use Sylius\Bundle\ApiBundle\Attribute\OrderTokenValueAware;
use Sylius\Bundle\ApiBundle\Attribute\PaymentIdAware;
use Sylius\Bundle\ApiBundle\Command\IriToIdentifierConversionAwareInterface;

/** @experimental */
#[OrderTokenValueAware]
#[PaymentIdAware]
class AddPaymentRequest implements IriToIdentifierConversionAwareInterface
```

To complete the suggestion of @NoResponseMate just bellow.

**Review by Prometee — DISMISSED** [review]

File: UPGRADE-2.0.md (hunk 1 above)
Comment (GSadee) [reply UPGRADE-2.0.md:11]: Directly no, but it was added on this branch, so there is no point in reversing it 🖖🏻

File: config/routes/test_cached_payum/routing.yaml (hunk 2 above)
Comment (GSadee) [reply config/routes/test_cached_payum/routing.yaml:5]: I guess so, it is added also in other test envs and if I see correctly, it is used for some Behat scenarios for theming 🤔

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/PaymentRequest.xml (hunk 16 above)
Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/Resources/config/serialization/PaymentRequest.xml:21]: It seems to be needed, otherwise we get:
```
request.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "Invalid identifier value or configuration." at ReadListener.php line 89 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): Invalid identifier value or configuration. at /Users/gsadee/Sites/Sylius/Sylius/vendor/api-platform/core/src/Symfony/EventListener/ReadListener.php:89)\n[previous exception] [object] (ApiPlatform\\Metadata\\Exception\\InvalidIdentifierException(code: 0): Parameter \"id\" not found, check the identifiers configuration. at /Users/gsadee/Sites/Sylius/Sylius/vendor/api-platform/core/src/State/UriVariablesResolverTrait.php:41)"} []
```

File: src/Sylius/Bundle/ApiBundle/Command/Payment/AddPaymentRequest.php (hunk 28 above)
Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/Command/Payment/AddPaymentRequest.php:21]: Adding this makes no sense as it would not be used at all, as we do not have payment id set as a parameter in the URL 🖖🏻

File: src/Sylius/Bundle/PaymentBundle/Provider/AbstractServiceProvider.php (hunk 27 above)
Comment (GSadee) [reply src/Sylius/Bundle/PaymentBundle/Provider/AbstractServiceProvider.php:23]: We will have to rethink this service and its implementation and probably refactor it, but after the merge of this branch into 2.0 🖖🏻

File: src/Sylius/Behat/Page/Admin/Payment/IndexPage.php (hunk 9 above)
Comment (GSadee) [reply src/Sylius/Behat/Page/Admin/Payment/IndexPage.php:59]: I have left it as it is, because here the generic show action on the grid is used

File: features/shop/payment_request/calling_notify_for_a_payment_request.feature [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "PHP T-Shirt" product
+        And the store ships everywhere for Free
+        And the store has a payment method "Offline" with a code "offline"
+        And there is a customer "john@example.com" that placed order with "PHP T-Shirt" product to "United States" based billing address with "Free" shipping method and "Offline" payment method
+
+
+    @ui
```
Comment (Rafikooo) [inline features/shop/payment_request/calling_notify_for_a_payment_request.feature:15]: The @api / @no-api tag should be added

**Review by Rafikooo — APPROVED** [review]

File: features/shop/payment_request/calling_notify_for_a_payment_request.feature (hunk 29 above)
Comment (GSadee) [reply features/shop/payment_request/calling_notify_for_a_payment_request.feature:15]: https://github.com/Sylius/Sylius/pull/17151#discussion_r1796888567
